### ISSUE TITLE
Duplicate sensor events

### ISSUE THREAD

**Issue by SuruPa00** [BIDMCDigitalPsychiatry/LAMP-platform#322]
> In accelerometer and gps (and potentially other sensor modalities as well) there are duplicate sensor events. The number of duplicates seems to vary by person. For some participants, there are a large number of duplicates which can hinder analysis. 

> This issue seems to be recent. Looking at an internal account data data, we see duplicates from June:

> 
<img width="727" alt="Screen_Shot_2021-07-19_at_10 17 32_AM" src="https://user-images.githubusercontent.com/86117379/126186924-36520c9b-e77b-4a95-8569-3eded3bae6c2.png">

> Please note that the test account has since been cleared of the duplicates in the database.

**Comment by dcurrey88** [comment]
Code to get duplicates for a certain participant:
`replic = db.col.aggregate([
    {'$match':{
        '_parent':participant_id,
        'sensor':'some sensor'
        }
    },
    { "$group": {
        "_id": {"timestamp": "$timestamp"},
        'idsUnicos': {'$addToSet': '$_id'},
        "total": { "$sum": 1 }
        }
    },
    {'$match': {
        'total': {'$gt': 1}
        }
    },
], allowDiskUse=True)`

Code to delete duplicates:
`response = []
for doc in replic:
    del doc['idsUnicos'][0]
    for id in doc['idsUnicos']:
        response.append(id)
db.sensor_event.delete_many({"_id": {"$in": response}})`

**Comment by jijopulikkottil** [comment]
hope this issue is getting with iOS app users! please confirm

**Comment by dcurrey88** [comment]
> hope this issue is getting with iOS app users! please confirm

I am seeing duplicate events for both iOS and Android.

**Comment by SuruPa00** [comment]
@jijopulikkottil @ZCOEngineer any update on this issue?

**Comment by Linoy339** [comment]
@SuruPa00 . We were trying to reproduce the issue.
Can you please confirm the environment? Is it staging or production?

**Comment by dcurrey88** [comment]
> @SuruPa00 . We were trying to reproduce the issue.
> Can you please confirm the environment? Is it staging or production?

I have seen duplicates in both the staging and production databases.

**Comment by Linoy339** [comment]
Can  you provide  a participant id in staging which you got this issue?

**Comment by Linoy339** [comment]
Also In staging, we have tested the bulk data write for the sensor events with duplicated data. It might be the reason in staging. We have tested with a few of participants. One of them is U5440992922(I don't remember other participants).

But we need to check deeply for what happened in production.

**Comment by SuruPa00** [comment]
Try the participants with the following ids: U3952228134 and U1455941750

**Comment by Linoy339** [comment]
Sure. This is production. Right?

**Comment by SuruPa00** [comment]
Yes this is production.

**Comment by jijopulikkottil** [comment]
@SuruPa00 same timestamp can appear on different sensors. hope this is not our case. [Could you share some sensor data which you see the duplicates. Or the query you used to find the duplicate, so we can confirm]

**Comment by dcurrey88** [comment]
> @SuruPa00 same timestamp can appear on different sensors. hope this is not our case. [Could you share some sensor data which you see the duplicates. Or the query you used to find the duplicate, so we can confirm]

Code to get duplicates for a certain participant from the database collection:
```python
replic = db.sensor.aggregate([
    {'$match':{
        '_parent':participant_id,
        'sensor':'some sensor'
        }
    },
    { "$group": {
        "_id": {"timestamp": "$timestamp"},
        'idsUnicos': {'$addToSet': '$_id'},
        "total": { "$sum": 1 }
        }
    },
    {'$match': {
        'total': {'$gt': 1}
        }
    },
], allowDiskUse=True)
```

Put duplicates in an array:
```python
response = []
for doc in replic:
    del doc['idsUnicos'][0]
    for id in doc['idsUnicos']:
        response.append(id)
```
You can see this for gps for the participants @SuruPa00 listed in the comment above.

**Comment by SuruPa00** [comment]
@jijopulikkottil @Linoy339  any update on this issue?

**Comment by jijopulikkottil** [comment]
Hi @SuruPa00, see our findings: 
The duplicates data seems to be very old.
When we check with this API https://api.lamp.digital/participant/{participantid}/sensor_event, the last collected data is on Jan.
U3952228134 is iOS user and U1455941750 is Android user.

When we discuss this issue, there is a possibility to enter duplicate data into database with following scenario:
 if the server process the sensor data partly or fully and doesn't response a 200 http status to mobile app. In this case, mobile app will resend data. At server side, this issue can happen when it writes data frequently to db and which leads to a crash.

There is a less probability to happen this case now because we improved our server/db capacity.

However to solve this case, we are not inserting sensor data immediately to db when received from mobile app, instead it will cache first and then will write. But this workaround is implemented only in Staging server now. 

@Linoy339 please confirm the above.

**Comment by Linoy339** [comment]
@SuruPa00 @jijopulikkottil . This is right. The possibility which we see is the same as @jijopulikkottil mentioned above.
For resolving frequent database writes of sensor_event data, we have introduced the cache.
This  will be available in production when we make a release soon

**Comment by SuruPa00** [comment]
@Linoy339 can we get an estimate of when the next release to include this fix will be ready? and is there anything you are waiting on our team for regarding this?

**Comment by Linoy339** [comment]
@SuruPa00 . Yes, we were testing Notification functionality with LAMP-worker enabled([https://github.com/BIDMCDigitalPsychiatry/LAMP-platform/issues/150].(https://github.com/BIDMCDigitalPsychiatry/LAMP-platform/issues/150)). It was tested from our side and found working as expected.  As  per the discussion, we need you to verify this and can do the release asap

**Comment by SuruPa00** [comment]
@Linoy339 thanks for checking in, make the release asap. Thanks

**Comment by SuruPa00** [comment]
@Linoy339 was this released

**Comment by Linoy339** [comment]
Yes @SuruPa00 . This is included in the latest release 2021.8.30
